EQL: correct time accounting for an intermediary response

bpintea opened the pull request:
This corrects the accounting of answer time reported in "took" attribute
of an intermediary EQL async response. This has been reported as a
negative value, since the delta mixes wall-clock time with VM time.

(The confusion is favoured by Task#getStartTime() returning a wall-clock
time, while Task#getStartTimeNanos() VM's time.)

Closes #75825.

bpintea committed:
Correct time account. for an intermediary EQL resp.

This corrects the accounting of answer time reported in "took" attribute
of an intermediary EQL async response. This has been reported as a
negative value, since the delta mixes wall-clock time with VM time.

(The confusion is favoured by Task#getStartTime() returning a wall-clock
time, while Task#getStartTimeNanos() VM's time.)

imotov commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -95,7 +95,8 @@ public void execute(EqlSearchRequest request, EqlSearchTask task, ActionListener
     public EqlSearchResponse initialResponse(EqlSearchTask task) {
         return new EqlSearchResponse(EqlSearchResponse.Hits.EMPTY,
-            threadPool.relativeTimeInMillis() - task.getStartTime(), false, task.getExecutionId().getEncoded(), true, true);
+            TimeValue.nsecToMSec(task.getStartTimeNanos()) - threadPool.relativeTimeInMillis(), false, task.getExecutionId().getEncoded(),
```
getStartTimeNanos is obtained from System.nanoTime(). So, would something like `TimeValue.nsecToMSec(System.nanoTime() - task.getStartTimeNanos())` make more sense in this case?

bpintea replied:
Thanks, indeed, _that_ makes sense.

bpintea committed:
Address review comments

Correct the accounting as the diff between VM's time and task's creation
time.

bpintea committed:
Fix test to add equality to delta 0 comparison

If search task is set in under 1ms, the test will complete with an
accounting of 0ms.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/eql_async_took_negative

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

Luegg reviewed (approved):
lgtm 👍 

Just wondering whether we have a general preference for the JVM time? Otherwise, `System.currentTimeMillis() - task.getStartTime()` might also be an option?

bpintea committed:
Address review comments

Switch to millis math.

bpintea commented:
> Just wondering whether we have a general preference for the JVM time? Otherwise, System.currentTimeMillis() - task.getStartTime() might also be an option?

You're right, the original code was doing millis math in the first place. I guess nanos math makes sense to more accurately differentiate between 0 and 1ms, but don't think this is that relevant in this case. Applied your suggestion.

astefan reviewed (approved):
LGTM
This is fixing an annoying issue that I noticed myself as well. Thanks for addressing it.

bpintea commented:
@elasticmachine run elasticsearch-ci/part-2

imotov commented:
> Just wondering whether we have a general preference for the JVM time?

I don't think it is a general preference, but for calculating elapsed time we are trying to go with `System.nanoTime()` nowadays. There are a couple of reasons for that. `System.nanoTime()`  is monotonic (which exception of some [JVM bugs](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6458294)), while `System.currentTimeMillis()` is not by design. In other words, `System.currentTimeMillis() - task.getStartTime()` can be negative if time on the system clock is update by a user or an NTP server at just the right moment. On some system you can also expect `System.nanoTime()` to be more precise than `System.currentTimeMillis()`. I saw reports of the first having granularity of hundreds of nanoseconds, while the second only ticking every 10ms or even 50ms on some older Windows systems. One notable exception when we would want to use `System.currentTimeMillis()` for elapse time calculation is if we move the data between nodes. `System.nanoTime()` doesn't really mean anything outside of JVM where the start time was computed, so if you start an event on one node and finish on another, `System.currentTimeMillis()` has to be used (but then all bets regarding monotonicity are off anyway since we don't synchronize clocks).

bpintea committed:
Revert last change, back to nano timing

Swtich back to converted nanos-accounting.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/eql_async_took_negative

bpintea commented:
@elasticmachine run elasticsearch-ci/docs

bpintea commented:
Thanks for the explanations, @imotov, reverted to your proposal.

elasticsearchmachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.13](https://github.com/elastic/elasticsearch/pull/76000)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/76000">](https://github.com/elastic/elasticsearch/pull/76000) |
| ✅ |  [7.14](https://github.com/elastic/elasticsearch/pull/76001)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/76001">](https://github.com/elastic/elasticsearch/pull/76001) |
